File: PartVision/src/types/detection.ts
// Represents a single detected object returned by the AI model
export interface PartDetection {
  label: string;      // e.g., "FRONT_BUMPER", "HEADLIGHT_L"
  confidence: number; // e.g., 0.92 (92% certainty)
  x_min: number;      // Normalized 0.0 - 1.0 bounding box coordinates
  y_min: number;
  width: number;
  height: number;
  polygon?: { x: number; y: number }[]; // Normalized polygon points from segmentation mask
}

// The full JSON response wrapper received over WebSocket from FastAPI
export interface DetectionResponse {
  detections: PartDetection[];
  process_time_ms: number; // Backend latency measurement
}

File: PartVisionFlutter/lib/models/detection.dart
class Detection {
  final String label;
  final double confidence;
  final double xMin;
  final double yMin;
  final double width;
  final double height;
  final List<PolygonPoint> polygon;

  Detection({
    required this.label,
    required this.confidence,
    required this.xMin,
    required this.yMin,
    required this.width,
    required this.height,
    required this.polygon,
  });

  factory Detection.fromJson(Map<String, dynamic> json) {
    final polygonList = json['polygon'] as List<dynamic>? ?? [];
    return Detection(
      label: json['label'] as String? ?? 'UNKNOWN',
      confidence: (json['confidence'] as num?)?.toDouble() ?? 0.0,
      xMin: (json['x_min'] as num?)?.toDouble() ?? 0.0,
      yMin: (json['y_min'] as num?)?.toDouble() ?? 0.0,
      width: (json['width'] as num?)?.toDouble() ?? 0.0,
      height: (json['height'] as num?)?.toDouble() ?? 0.0,
      polygon: polygonList
          .map((p) => PolygonPoint(
                x: (p['x'] as num?)?.toDouble() ?? 0.0,
                y: (p['y'] as num?)?.toDouble() ?? 0.0,
              ))
          .toList(),
    );
  }
}

class PolygonPoint {
  final double x;
  final double y;
  const PolygonPoint({required this.x, required this.y});
}
